### ISSUE TITLE
Memory issues after finishing one epoch

### ISSUE THREAD

**Issue by helson73** [OpenNMT/OpenNMT-py#476]
Similar issue with #301 
During one epoch, memory usage is quite consistent, but after finishing one epoch, huge additional memory usage suddenly raised. 
Any idea what causes this?

Update: it only happens after the first epoch.

**Comment by helson73** [comment]
Test Code:
```python
    def validate(self):
        """ Called for each epoch to validate. """
        # Set model in validating mode.
        self.model.eval()

        stats = Statistics()

        print 'step --0'
        show_memusage()
        first = True
        for batch in self.valid_iter:
            src = onmt.io.make_features(batch, 'src', self.data_type)
            if self.data_type == 'text':
                _, src_lengths = batch.src
            else:
                src_lengths = None

            tgt = onmt.io.make_features(batch, 'tgt')
            if first:
                print 'step --1'
                show_memusage()

            # F-prop through the model.
            outputs, attns, _ = self.model(src, tgt, src_lengths)
            if first:
                print 'step --2'
                show_memusage()

            # Compute loss
            batch_stats = self.valid_loss.monolithic_compute_loss(
                    batch, outputs, attns)
            if first:
                print 'step --3'
                show_memusage()

            # Update statistics.
            stats.update(batch_stats)
            first = False

        # Set model back to training mode.
        self.model.train()

        return stats
```

**Comment by helson73** [comment]
Result:
**label_smoothing enabled:** 
Epoch  1,   100/  313; acc:  10.38; ppl: 5901.63; 6973 src tok/s; 7052 tgt tok/s;     16 s elapsed
Epoch  1,   200/  313; acc:  17.89; ppl: 957.51; 7023 src tok/s; 7114 tgt tok/s;     32 s elapsed
Epoch  1,   300/  313; acc:  22.84; ppl: 389.45; 7060 src tok/s; 7129 tgt tok/s;     48 s elapsed
Train perplexity: 1231.65
Train accuracy: 17.3245
step --0
VRAM usage: 5085
step --1
VRAM usage: 5085
step --2
VRAM usage: 5085
step --3
VRAM usage: 5541
Validation perplexity: 556.284
Validation accuracy: 18.7566

**label_smoothing disabled:**
Epoch  1,   100/  313; acc:   9.02; ppl: 6344.58; 7157 src tok/s; 7237 tgt tok/s;     16 s elapsed
Epoch  1,   200/  313; acc:  18.49; ppl: 975.34; 7167 src tok/s; 7239 tgt tok/s;     31 s elapsed
Epoch  1,   300/  313; acc:  22.92; ppl: 388.92; 7246 src tok/s; 7324 tgt tok/s;     46 s elapsed
Train perplexity: 1261.71
Train accuracy: 17.1174
step --0
VRAM usage: 5605
step --1
VRAM usage: 5605
step --2
VRAM usage: 5605
step --3
VRAM usage: 6061
Validation perplexity: 538.533
Validation accuracy: 17.876

It shows additional memory usage is raised by validate compute loss.

**Comment by srush** [comment]
Is this a new issue? I wonder if something is not being marked "violatile" during validation. That would cause it to hold onto a lot of objects for backprop.

**Comment by helson73** [comment]
@srush I think this is the same as #301 
I am new to pytorch so I am not sure whether this happens to all pytorch apps or just ours.

**Comment by srush** [comment]
Okay ill check it out soon.

**Comment by helson73** [comment]
@srush 
I think I found the reason.
If max_generator_batches is smaller than batch_size, generator first allocated a space (max_gen_batch, vocab_size) in training mode, then in validation mode, generator found that (valid_batch_size, vocab_size) exceeds pre-allocated space, so it allocated additional spaces.
When I set max_generator_batches == batch_size, I got following result:

Epoch  1,   100/  313; acc:   9.41; ppl: 5808.32; 6995 src tok/s; 7084 tgt tok/s;     16 s elapsed
Epoch  1,   200/  313; acc:  18.53; ppl: 968.56; 7218 src tok/s; 7296 tgt tok/s;     31 s elapsed
Epoch  1,   300/  313; acc:  22.92; ppl: 399.18; 7140 src tok/s; 7203 tgt tok/s;     47 s elapsed
Train perplexity: 1228.31
Train accuracy: 17.2972
before validate -- VRAM usage: 6385
after validate -- VRAM usage: 6385
Validation perplexity: 551.711
Validation accuracy: 19.091

I think it's better to implement validate in shards.

**Comment by srush** [comment]
Let's keep this issue open until we merge. I'm not sure I completely understand it.

**Comment by helson73** [comment]
Sorry, I misunderstood the meaning of max_gen_batches.

First, I don't think this problem is caused by unnecessary gradients.
I checked out which part caused problem, and found it was generator(marked as eval in validation mode).
Also, the input of generator, found it was marked as "volatile" in validation mode.

I think the problem is caused by pre-allocated space of generator.
If I am right, every node in computation graph has caches for its inputs & outputs. The size of cache depends on shapes of input & output tensors. If the node encounters tensor size exceeds pre-allocated space, then it just re-allocate new space for caches.

In this case:

During training, the output of generator has following shapes.
(max_gen_batches, batch_size, vocab_size)
The generator will re-use space of output until new ones exceed that size.

During validation, the output of generator has following shapes.
(max_valid_tgt_length, valid_batch_size, vocab_size)
If max_valid_tgt_length * valid_batch_size > max_gen_batches * batch_size, generator have to allocate additional space.

Currently, valid_batch_size == batch_size, therefore if max_valid_tgt_length > max_gen_batches, problem happens.

For example, let max_gen_batches = 32, batch_size = 64.
If any sentence in validset has length longer than 32, the generator have to re-allocate new space to store the output.

When I set valid_batch_size to 32 (because misunderstood meaning of max_gen_batches),
max_gen_batches * batch_size = 32 * 64 = 2048 > max_valid_tgt_length * valid_batch_size = 1600
Pre-allocated space is larger than space needed in validation, therefore no additional allocation occurred.
That explains results in my comment above.
